### PR TITLE
[WIP] Added testing

### PR DESCRIPTION
**Pull request by waltsims**
Add extensive testing against k-Wave functions

### CODE REVIEW

File: tests/matlab_test_data_collectors/matlab_collectors/collect_makeLine.m [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+% line = makeLine(10, 20, [4, 15], [10, 1]);
+
+% line = makeLine(20, 20, [15, 15], 0.75 * pi, 5);
+% 
+% line = makeLine(20, 20, [15, 15], 0.75 * pi, -5);
+% line = makeLine(20, 10, [15, 10], 0.1 * pi, 5);
+% 
+% figure;
+% imagesc(line);
+
```
Comment (waltsims) [inline tests/matlab_test_data_collectors/matlab_collectors/collect_makeLine.m:14]: clean up

File: tests/matlab_test_data_collectors/matlab_collectors/collect_reorderSensorData.m [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+mask = zeros(mask_size(1), mask_size(2));
+
+for i=1:20
+    mask(randi(mask_size(1)), randi(mask_size(2))) = 1;
+end
+sensor = struct('mask', mask);
+
+% sensor_data = rand(sensor_data_size);
+
```
Comment (waltsims) [inline tests/matlab_test_data_collectors/matlab_collectors/collect_reorderSensorData.m:14]: clean up

File: tests/matlab_test_data_collectors/python_testers/cart2grid_test.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+from kwave.utils.conversionutils import scale_time
+
+from scipy.io import loadmat
+import numpy as np
+import os
+import pytest
+
+
+@pytest.mark.skip(reason="Missing cart2grid collector")
```
Comment (waltsims) [inline tests/matlab_test_data_collectors/python_testers/cart2grid_test.py:13]: @faridyagubbayli add car2grid collector and remove this skip

File: tests/test_filterutils.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                                   0.24197072, 0.22988214, 0.21785218, 0.20593627, 0.19418605,
+                                   0.18264909, 0.17136859, 0.16038333, 0.14972747, 0.13943057,
+                                   0.1295176, 0.120009, 0.11092083, 0.10226492, 0.09404908,
+                                   0.08627732, 0.07895016, 0.07206487, 0.06561581, 0.05959471,
+                                   0.05399097, 0.04879202, 0.0439836, 0.03955004, 0.03547459,
+                                   0.03173965, 0.02832704, 0.02521822, 0.02239453, 0.01983735,
+                                   0.0175283, 0.01544935, 0.01358297, 0.01191224, 0.01042093,
+                                   0.00909356, 0.00791545, 0.00687277, 0.00595253, 0.00514264,
+                                   0.00443185]])
+    assert (expected_gaussian - gauss_distr < 1e-6).all(), "Gaussian distribution did not match expected distribution"
```
Comment (waltsims) [inline tests/test_filterutils.py:37]: Test with a collector @waltsims

File: tests/test_filterutils.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                               6.23085477e-215, 5.22990800e-215, 4.02797880e-215,
+                               2.67125663e-215, 1.21187957e-215, -2.94069337e-216,
+                               -1.78871733e-215, -3.21462592e-215, -4.51699824e-215,
+                               -5.64578484e-215, -6.55760711e-215, -7.21742421e-215,
+                               -7.59987974e-215, -7.69027616e-215, -7.48513958e-215,
+                               -6.99235328e-215, -6.23085477e-215]])
+    fs = 512  # [Hz]
+    data = generate_test_signal(fs)
+    data = gaussian_filter(data, fs, 15, 0.5)
+    assert (abs(expected_data - data) < 1e-3).all()
```
Comment (waltsims) [inline tests/test_filterutils.py:95]: test with collector @waltsims

File: tests/test_filterutils.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+def test_smooth():
+    A = np.diag(np.ones(6))
+    A_sm = np.array(smooth(A))
+    expected_A_sm = np.array([[0.3150659, 0.23343146, 0.09246705, 0.03313708, 0.09246705, 0.23343146],
+                     [0.23343146, 0.3150659, 0.23343146, 0.09246705, 0.03313708, 0.09246705],
+                     [0.09246705, 0.23343146, 0.3150659, 0.23343146, 0.09246705, 0.03313708],
+                     [0.03313708, 0.09246705, 0.23343146, 0.3150659, 0.23343146, 0.09246705],
+                     [0.09246705, 0.03313708, 0.09246705, 0.23343146, 0.3150659, 0.23343146],
+                     [0.23343146, 0.09246705, 0.03313708, 0.09246705, 0.23343146, 0.3150659]])
+    assert (A_sm - expected_A_sm < 0.001).all()
```
Comment (waltsims) [inline tests/test_filterutils.py:174]: test with collector @waltsims

File: tests/test_kutils.py [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+                                   0.99017714, 0.99992254, 1., 1., 1.,
+                                   1., 1., 1., 1., 1.,
+                                   1., 1., 1., 1., 1.,
+                                   1., 1., 1., 1., 1.,
+                                   1., 1., 1., 1., 1.,
+                                   1., 1., 1., 1., 1.,
+                                   1., 1., 0.99992254, 0.99017714, 0.9616098,
+                                   0.91531474, 0.85306778, 0.7772581, 0.69079646, 0.59700331,
+                                   0.4994812, 0.40197633, 0.30823458, 0.22185752, 0.14616399,
+                                   0.08406256, 0.03793968, 0.00956799, 0.])).all()
```
Comment (waltsims) [inline tests/test_kutils.py:97]: rewrite with collector @waltsims

File: tests/test_misc.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    # evaluate pressure
+    [p_axial, p_lateral] = focused_bowl_oneil(radius, diameter,
+                                              velocity, frequency, sound_speed, density,
+                                              axial_position, lateral_position)
+    # compare first and last vals
+    assert np.isclose(p_axial[0], 1227.01736354) and np.isclose(p_axial[-1], 13036.24465513)
+    assert np.isclose(p_lateral[0], 82805.42312207383) and np.isclose(p_lateral[-1], 82805.42312207383)
+    assert np.isclose(p_lateral[5], 12880.47179330867)  # one final test value
+
```
Comment (waltsims) [inline tests/test_misc.py:50]: test with collector @waltsims

File: tests/test_pmlutils.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    # time array
+    dt = 2e-9  # [s]
+    t_end = 300e-9  # [s]
+    # create the computational grid
+    kgrid = kWaveGrid([nx, nx, nx], [dx, dx, dx])
+
+    # create the time array
+    kgrid.setTime(round(t_end / dt) + 1, dt)
+
+    assert (32 == getOptimalPMLSize(kgrid)).all()
```
Comment (waltsims) [inline tests/test_pmlutils.py:58]: test with collector @waltsims

File: tests/test_reconstruction.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                                           -1.62550429e-01, -3.13278219e-01, -4.26343806e-01, -4.89896528e-01,
+                                           -5.00386632e-01, -4.63525010e-01, -3.91742171e-01, -2.99906977e-01,
+                                           -2.01695338e-01, -1.07714330e-01, -2.51323109e-02, 4.18960027e-02,
+                                           9.15678839e-02, 1.23813253e-01, 1.39803369e-01, 1.41581150e-01,
+                                           1.31776163e-01, 1.13365585e-01, 8.94527423e-02, 6.30499609e-02,
+                                           3.68688594e-02, 1.31353908e-02, -6.54543545e-03, -2.12524482e-02,
+                                           -3.07282290e-02, -3.52829902e-02, -3.56452574e-02, -3.27905068e-02,
+                                           -2.77777264e-02, -2.16155965e-02, -1.21840122e-02, -2.53226773e-03,
+                                           0.00000000e+00])
+
```
Comment (waltsims) [inline tests/test_reconstruction.py:58]: test with collector @waltsims

File: tests/test_water.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+                       5.79006441, 5.80461054, 5.81903673, 5.83333264, 5.84748682,
+                       5.8614867, 5.87531864, 5.88896788, 5.90241855, 5.9156537,
+                       5.92865527, 5.9414041, 5.95387992, 5.96606138, 5.977926,
+                       5.98945022, 6.00060938, 6.0113777, 6.02172833, 6.03163328,
+                       6.04106348, 6.04998878, 6.05837788, 6.06619841, 6.07341691,
+                       6.0799988, 6.08590839, 6.0911089, 6.09556247, 6.0992301,
+                       6.10207171, 6.10404612, 6.10511105, 6.10522312, 6.10433782,
+                       6.10240959]
+    for idx, value in enumerate(expected_values):
+        assert abs(value - water_non_linearity(idx)) < 1e-6, "Expected value deviates from water_non_linearity output"
```
Comment (waltsims) [inline tests/test_water.py:114]: Test with collector @waltsims

File: tests/test_filterutils.py (hunk 5 above)
Comment (waltsims) [reply tests/test_filterutils.py:95]: done

**Review by faridyagubbayli — APPROVED** [review]
Looks great :)

**Pull request merged**